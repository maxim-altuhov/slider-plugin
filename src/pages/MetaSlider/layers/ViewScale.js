import $ from 'jquery';
import Observer from '../patterns/Observer';
class ViewScale extends Observer {
  constructor() {
    super();
    this.isFirstInit = true;
  }

  // Первоначальная инициализация
  init(options) {
    this.$selector = options.$selector;
    this.$elemSlider = options.$elemSlider;
    this.$elemScale = options.$elemScale;
  }

  // Обновление view
  update(options) {
    if (this.isFirstInit) {
      this.init(options);
      this.isFirstInit = false;
    }

    const { key } = options;
    const renderScaleVerifKeys = (
      key === 'init'
      || key === 'showScale'
      || key === 'initAutoScaleCreation'
      || key === 'step'
      || key === 'stepSizeForScale'
      || key === 'minValue'
      || key === 'maxValue'
      || key === 'calcNumberOfDecimalPlaces'
      || key === 'numberOfDecimalPlaces'
      || key === 'customValues'
      || key === 'initFormatted'
      || key === 'preFix'
      || key === 'postFix'
      || key === 'initScaleAdjustment'
    );
    const styleVerifKeys = (
      key === 'init'
      || key === 'colorForScale'
      || key === 'showScale'
    );

    if (renderScaleVerifKeys) {
      this.createScale(options);
      this.checkingScaleSize(options);

      if (key === 'initScaleAdjustment' || key === 'init') this.setEventsWindow(options);
    }
    if (styleVerifKeys) this.setStyleForScale(options);
  }

  // Рендер шкалы значений
  createScale(options) {
    if (options.showScale) {
      this.$elemScale.empty();

      const {
        initAutoScaleCreation,
        step,
        stepSizeForScale,
        minValue,
        maxValue,
        customValues,
        initFormatted,
        preFix,
        postFix,
      } = options;

      const $fragmentWithScale = $(document.createDocumentFragment());
      const stepSizeValue = initAutoScaleCreation ? step : stepSizeForScale;
      let currentValue = minValue;

      for (; currentValue <= maxValue; currentValue += stepSizeValue) {
        const isCustomValue = customValues.length > 0;
        const convertedValue = initFormatted ? currentValue.toLocaleString() : currentValue;
        const resultValue = isCustomValue ? customValues[currentValue] : convertedValue;

        const elemScalePoint = `<button type="button" class="meta-slider__scale-point js-meta-slider__scale-point" style="color: inherit" data-value="${currentValue}">${preFix}${resultValue}${postFix}</button>`;

        $fragmentWithScale.append(elemScalePoint);
      }

      this.$elemScale.append($fragmentWithScale);

      this.$elemScalePoints = this.$selector.find('.js-meta-slider__scale-point');
      this.scalePointsSize = 0;

      this.$elemScalePoints.each((index, scalePoint) => {
        const $scalePoint = $(scalePoint);
        const valueInScalePoint = Number($scalePoint.attr('data-value'));

        const resultValue = (valueInScalePoint - minValue) / (maxValue - minValue);
        this.scalePointsSize += $scalePoint.outerWidth();

        $scalePoint.css('left', `${resultValue * 100}%`);
      });

      if (this.mapSkipScalePoints && this.mapSkipScalePoints.size > 0) {
        this.mapSkipScalePoints.clear();
      } else {
        this.mapSkipScalePoints = new Map();
      }

      this.setEventsScalePoints();
    }
  }

  // Устанавливаем стили для шкалы
  setStyleForScale(options) {
    const { colorForScale, showScale } = options;

    if (showScale) {
      this.$elemScale.css({
        borderColor: colorForScale,
        color: colorForScale,
        opacity: 1,
        'pointer-events': '',
      });
      this.$elemScale.children().removeAttr('tabindex');
    } else {
      this.$elemScale.css({ opacity: 0, 'pointer-events': 'none' });
      this.$elemScale.children().attr('tabindex', -1);
    }
  }

  /**
   * Метод, который проверяет помещаются ли все деления шкалы на данной длине слайдера.
   * Если нет, то этот метод делает авто-подстройку делений шкалы и скрывает лишние значения
   */
  checkingScaleSize(options) {
    const { showScale, initScaleAdjustment } = options;

    if (showScale && initScaleAdjustment) {
      const MARGIN_PX = 100;
      const sliderSize = this.$elemSlider.outerWidth();

      while (this.scalePointsSize + MARGIN_PX > sliderSize) {
        const totalSizeScalePoints = this.scalePointsSize + MARGIN_PX;
        this.skipScalePointsArray = [];
        this.$elemScalePoints = this.$selector.find('.js-meta-slider__scale-point:not(.meta-slider__scale-point_skip)');
        this.scalePointsSize = 0;

        const sizeScalePointsArray = this.$elemScalePoints.length;

        if (sizeScalePointsArray <= 2) break;

        this.$elemScalePoints.each((index, scalePoint) => {
          const $currentScalePoint = $(scalePoint);
          const firstOrLastIndex = (index === 0) || (index === (sizeScalePointsArray - 1));
          const intervalWithoutFirstAndLastIndex = !firstOrLastIndex && sizeScalePointsArray <= 6;

          if (index % 2 !== 0 && sizeScalePointsArray > 6) {
            this.setPropForSkipScalePoint($currentScalePoint);
          } else if ((sizeScalePointsArray % 2 !== 0) && sizeScalePointsArray <= 6) {
            this.setPropForSkipScalePoint($currentScalePoint);
          } else if ((sizeScalePointsArray % 2 === 0) && intervalWithoutFirstAndLastIndex) {
            this.setPropForSkipScalePoint($currentScalePoint);
          }

          if (!$currentScalePoint.hasClass('meta-slider__scale-point_skip')) this.scalePointsSize += $currentScalePoint.outerWidth();
        });

        this.mapSkipScalePoints.set(totalSizeScalePoints, [...this.skipScalePointsArray]);
      }

      this.checkingSkipScalePointSize(sliderSize, MARGIN_PX);
    }
  }

  // Установка стилей для пропущенных делений шкалы
  setPropForSkipScalePoint($scalePoint) {
    $scalePoint.addClass('meta-slider__scale-point_skip')
      .attr('tabindex', -1)
      .css({ color: 'transparent', borderColor: 'inherit' });

    this.skipScalePointsArray.push($scalePoint);
  }

  /**
   * Метод отслеживает размер слайдера и возвращает скрытые деления шкалы,
   * если они уже помещаются на шкале
   */
  checkingSkipScalePointSize(sliderSize, margin) {
    this.mapSkipScalePoints.forEach((scalePointSkipArray, controlSize) => {
      if (sliderSize > controlSize + (margin / 3)) {
        scalePointSkipArray.forEach(($scalePoint) => {
          $scalePoint.removeAttr('tabindex')
            .removeClass('meta-slider__scale-point_skip')
            .css({ color: 'inherit', borderColor: '' });

          this.scalePointsSize += $scalePoint.outerWidth();
        });

        this.mapSkipScalePoints.delete(controlSize);
      }
    });
  }

  // Обработчик события отслеживающий размер окна браузера для метода checkingScaleSize()
  setEventsWindow(options) {
    const { showScale, initScaleAdjustment } = options;

    if (showScale && initScaleAdjustment) {
      $(window).on('resize.scale', this.handleCheckingScaleSize.bind(this, options));
    } else {
      $(window).off('resize.scale');
    }
  }

  // Отслеживает ширину слайдера
  handleCheckingScaleSize(options) {
    this.checkingScaleSize(options);
  }

  // Обрабочик событий кликов на значения шкалы
  setEventsScalePoints() {
    this.$elemScalePoints.each((index, elemPoint) => {
      $(elemPoint).on('click.scalePoint', this.handleGetValueInScalePoint.bind(this));
    });
  }

  // Получает значения при клике на шкалу слайдера
  handleGetValueInScalePoint(event) {
    event.preventDefault();
    const $target = $(event.target);
    let targetValue = Number($target.attr('data-value'));

    this.notify(event, targetValue);
  }
}

export default ViewScale;
