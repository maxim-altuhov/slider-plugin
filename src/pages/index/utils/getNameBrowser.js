function getNameBrowser() {
  const ua = navigator.userAgent;

  if (ua.search(/rv:11.0/) > 0) {
    return 'Internet Explorer 11';
  }
  if (ua.search(/MSIE/) > 0) {
    return 'Internet Explorer';
  }
  return 'ะะต Internet Explorer';
}

export default getNameBrowser;
