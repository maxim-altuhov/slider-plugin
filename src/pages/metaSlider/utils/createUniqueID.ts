function createUniqueID(): string {
  const IDPartOne = Date.now().toString(36);
  const IDPartTwo = Math.random().toString(36).substring(2);

  return IDPartOne + IDPartTwo;
}

export default createUniqueID;
