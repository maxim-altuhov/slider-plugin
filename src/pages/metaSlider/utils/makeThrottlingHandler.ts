// The any type is used, since the called functions can contain any arguments
function makeThrottlingHandler(fn: (...args: any[]) => unknown, timeout: number) {
  let timer: NodeJS.Timeout | null = null;

  return (...args: any[]) => {
    if (timer) return;

    timer = setTimeout(() => {
      fn(...args);

      if (timer) {
        clearTimeout(timer);
        timer = null;
      }
    }, timeout);
  };
}

export default makeThrottlingHandler;
