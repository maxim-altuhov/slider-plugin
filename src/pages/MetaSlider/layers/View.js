import ViewSlider from './ViewSlider';
import ViewScale from './ViewScale';
import ViewMarkers from './ViewMarkers';
import ViewThumbs from './ViewThumbs';
import ViewMinAndMaxValues from './ViewMinAndMaxValues';
import ViewError from './ViewError';

class View {
  constructor() {
    this.views = {
      viewError: new ViewError(),
      viewSlider: new ViewSlider(),
      viewThumbs: new ViewThumbs(),
      viewMarkers: new ViewMarkers(),
      viewMinAndMax: new ViewMinAndMaxValues(),
      viewScale: new ViewScale(),
    };
  }

  init(options) {
    Object.keys(this.views).forEach((view) => {
      if ('init' in this.views[view]) this.views[view].init(options);
    });
  }

  renderSlider(initSelector) {
    this.views.viewSlider.renderSlider(initSelector);
  }
}

export default View;
