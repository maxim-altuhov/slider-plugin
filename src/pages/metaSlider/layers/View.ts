import ViewSlider from './ViewSlider';
import ViewScale from './ViewScale';
import ViewMarkers from './ViewMarkers';
import ViewThumbs from './ViewThumbs';
import ViewError from './ViewError';

class View {
  /**
   * I use the any type to be able to access any method
   * of the class instance in the viewList object.
   */
  readonly viewList: Record<string, any> = {
    viewError: new ViewError(),
    viewScale: new ViewScale(),
    viewSlider: new ViewSlider(),
    viewThumbs: new ViewThumbs(),
    viewMarkers: new ViewMarkers(),
  };

  update(options: IPluginOptions) {
    Object.keys(this.viewList).forEach((view) => {
      if ('update' in this.viewList[view]) this.viewList[view].update(options);
    });
  }

  // Initial render of the slider and its main elements
  renderSlider($initSelector: JQuery) {
    this.viewList['viewSlider'].renderSlider($initSelector);
  }
}

export default View;
