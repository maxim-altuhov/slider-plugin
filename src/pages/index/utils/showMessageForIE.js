import getNameBrowser from './getNameBrowser';

function showMessageForIE() {
  const currentBrowser = getNameBrowser();
  const thisIsIEBrowser = (currentBrowser === 'Internet Explorer 11') || (currentBrowser === 'Internet Explorer');

  if (thisIsIEBrowser) {
    document.body.innerHTML = 'Internet Explorer is no longer supported, please switch to a more modern browser!';
  }
}

export default showMessageForIE;
