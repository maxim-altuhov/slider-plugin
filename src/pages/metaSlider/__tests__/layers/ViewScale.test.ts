/* eslint-disable @typescript-eslint/dot-notation */
// The tests uses the 'any' type in jest.spyOn so that private methods of the class can be tested
import View from '../../layers/View';
import ViewScale from '../../layers/ViewScale';
import InitSettings from '../../data/InitSettings';
import * as makeThrottlingHandler from '../../utils/makeThrottlingHandler';

jest.mock('../../utils/createUniqueID');
const classMainView = new View();
document.body.innerHTML = '<div id="render-selector"></div>';
const $initSelector = $('#render-selector');
classMainView.renderSlider($initSelector);

describe('Checking the "ViewScale" layer, before first initialization.', () => {
  const notInitViewScale = new ViewScale();

  test('State before first initialization the "update" method', () => {
    expect($initSelector.html()).toMatchSnapshot();
    expect(notInitViewScale['_$selector']).toHaveLength(0);
    expect(notInitViewScale['_$elemSlider']).toHaveLength(0);
    expect(notInitViewScale['_$elemScale']).toHaveLength(0);
    expect(notInitViewScale['_$elemScalePoints']).toHaveLength(0);
    expect(notInitViewScale['_scalePointsSize']).toBe(0);
    expect(notInitViewScale['_mapSkipScalePoints'].size).toBe(0);
    expect(notInitViewScale['_skipScalePointsArray']).toHaveLength(0);
    expect(notInitViewScale['_isFirstInit']).toBe(true);
    expect(notInitViewScale.observerList).toHaveLength(0);
    expect(notInitViewScale['_verifKeysObj']).toBeDefined();
  });
});

describe('Checking the "ViewScale" layer', () => {
  const SLIDER_SIZE_PX = 250;
  const classViewScale = new ViewScale();
  const defaultSettings = {
    key: 'init',
    $selector: $initSelector,
    $elemSlider: $('.js-meta-slider'),
    $elemScale: $('.js-meta-slider__scale'),
    showScale: true,
    initAutoScaleCreation: true,
    initFormatted: false,
    initScaleAdjustment: true,
    step: 1,
    stepSizeForScale: 2.5,
    minValue: 1,
    maxValue: 5,
    customValues: [],
    preFix: '',
    postFix: '',
    colorForScale: 'black',
  };

  let testSettings: IPluginOptions;
  let $selector: JQuery<HTMLElement>;
  let $elemSlider: JQuery<HTMLElement>;
  let $elemScale: JQuery<HTMLElement>;
  let sliderID: string | undefined;

  beforeAll(() => {
    testSettings = $.extend({}, InitSettings, defaultSettings);
    classViewScale.update(testSettings);
    $selector = classViewScale['_$selector'];
    $elemSlider = classViewScale['_$elemSlider'];
    $elemScale = classViewScale['_$elemScale'];
    sliderID = $selector.attr('data-id');

    $elemSlider.css('width', `${SLIDER_SIZE_PX}px`);
  });

  afterEach(() => {
    testSettings = $.extend({}, InitSettings, defaultSettings);
    jest.restoreAllMocks();
  });

  test('Checking the "_init" method and the first initialization update method', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_setEventResize');

    classViewScale['_isFirstInit'] = true;
    classViewScale.update(testSettings);

    expect(classViewScale['_$selector']).toHaveLength(1);
    expect(classViewScale['_$elemSlider']).toHaveLength(1);
    expect(classViewScale['_$elemScale']).toHaveLength(1);
    expect(classViewScale['_isFirstInit']).toBe(false);
    expect(classViewScale['_setEventResize']).toHaveBeenCalledWith(testSettings);
  });

  test('Checking the "_createScale" method => "defaultSettings"', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_setEventsScalePoints');
    jest.spyOn<ViewScale, any>(classViewScale, '_calcScalePointsSize');
    jest.spyOn(Map.prototype, 'clear');
    jest.spyOn($.fn, 'empty');
    const CONTROL_SIZE_PX = 100;
    const { maxValue, minValue, step } = testSettings;

    expect(classViewScale['_mapSkipScalePoints'].size).toBe(0);
    classViewScale['_mapSkipScalePoints'].set(CONTROL_SIZE_PX, [
      classViewScale['_$elemScalePoints'].eq(0),
    ]);
    expect(classViewScale['_mapSkipScalePoints'].size).not.toBe(0);
    classViewScale.update(testSettings);

    const $elemScalePoints = classViewScale['_$elemScalePoints'];
    let scalePointCounter = 0;

    for (let currentValue = minValue; currentValue <= maxValue; currentValue += step) {
      expect($elemScalePoints.eq(scalePointCounter).attr('data-value')).toBe(String(currentValue));
      expect($elemScalePoints.eq(scalePointCounter).text()).toBe(String(currentValue));

      scalePointCounter += 1;
    }
    expect(classViewScale['_$elemScale'].html()).toMatchSnapshot();
    expect(classViewScale['_$elemScale'].empty).toHaveBeenCalled();
    expect($elemScalePoints).toHaveLength(scalePointCounter);

    $elemScalePoints.each((_, scalePoint) => {
      const $scalePoint = $(scalePoint);
      expect($scalePoint.attr('class')).toMatch(/js-meta-slider__scale-point/);
      expect($scalePoint.attr('data-value')).toBeDefined();
      expect($scalePoint.prop('style')).toHaveProperty('color', '');
      expect($scalePoint.prop('style')).toHaveProperty(
        'left',
        `${((Number($scalePoint.attr('data-value')) - minValue) / (maxValue - minValue)) * 100}%`,
      );
      expect(classViewScale['_calcScalePointsSize']).toHaveBeenCalledWith(scalePoint, testSettings);
    });

    expect(classViewScale['_setEventsScalePoints']).toHaveBeenCalled();
    expect(classViewScale['_mapSkipScalePoints'].clear).toHaveBeenCalled();
    expect(classViewScale['_mapSkipScalePoints'].size).toBe(0);
  });

  test('Checking the "_createScale" method => option "preFix/postFix"', () => {
    testSettings.key = 'preFix';
    testSettings.preFix = 'abc';
    testSettings.postFix = 'cba';
    classViewScale.update(testSettings);
    const $elemScalePoints = classViewScale['_$elemScalePoints'];

    $elemScalePoints.each((_, scalePoint) => expect(scalePoint.textContent).toMatch(/abc\d+cba/));
  });

  test('Checking the "_createScale" method => option "initAutoScaleCreation"', () => {
    testSettings.key = 'initAutoScaleCreation';
    testSettings.initAutoScaleCreation = false;
    testSettings.stepSizeForScale = null;
    testSettings.step = 1;
    const { maxValue, minValue } = testSettings;
    classViewScale.update(testSettings);

    const $elemScalePoints = classViewScale['_$elemScalePoints'];
    let scalePointCounter = 0;

    for (let currentValue = minValue; currentValue <= maxValue; currentValue += testSettings.step) {
      expect($elemScalePoints.eq(scalePointCounter).attr('data-value')).toBe(String(currentValue));
      expect($elemScalePoints.eq(scalePointCounter).text()).toBe(String(currentValue));

      scalePointCounter += 1;
    }
  });

  test('Checking the "_createScale" method => option "initFormatted"', () => {
    testSettings.key = 'initFormatted';
    testSettings.initFormatted = true;
    testSettings.minValue = 1000;
    testSettings.maxValue = 10000;
    testSettings.step = 1000;
    classViewScale.update(testSettings);

    const $elemScalePoints = classViewScale['_$elemScalePoints'];

    $elemScalePoints.each((_, scalePoint) => {
      const $scalePoint = $(scalePoint);
      expect($scalePoint.text()).toBe(Number($scalePoint.attr('data-value')).toLocaleString());
    });
  });

  test('Checking the "_createScale" method => option "customValues"', () => {
    testSettings.key = 'customValues';
    testSettings.customValues = ['a', 'b', 'c', 'd'];
    const { customValues } = testSettings;

    testSettings.minValue = 0;
    testSettings.maxValue = customValues.length - 1;
    testSettings.initAutoScaleCreation = false;
    testSettings.checkingStepSizeForScale = false;
    testSettings.initFormatted = false;
    testSettings.calcNumberOfDecimalPlaces = false;
    testSettings.numberOfDecimalPlaces = 0;
    testSettings.step = 1;
    testSettings.stepSizeForScale = 1;
    classViewScale.update(testSettings);

    const $elemScalePoints = classViewScale['_$elemScalePoints'];

    $elemScalePoints.each((index, scalePoint) => {
      const $scalePoint = $(scalePoint);
      expect($scalePoint.text()).toBe(customValues[index]);
    });
  });

  test.each`
    isVerticalStatus | testScalePointHeight | testScalePointWidth
    ${true}          | ${50}                | ${null}
    ${false}         | ${null}              | ${100}
    ${true}          | ${null}              | ${null}
    ${false}         | ${null}              | ${null}
  `(
    'Checking the "_calcScalePointsSize" method => option "isVertical"',
    ({ isVerticalStatus, testScalePointHeight, testScalePointWidth }) => {
      const $testScalePoint = classViewScale['_$elemScalePoints'].eq(0);

      testSettings.isVertical = isVerticalStatus;
      testSettings.testHeight = testScalePointHeight;
      testSettings.testWidth = testScalePointWidth;
      classViewScale['_calcScalePointsSize']($testScalePoint[0], testSettings);

      const { isVertical, testHeight, testWidth } = testSettings;

      if (isVertical && testHeight) {
        expect(classViewScale['_scalePointsSize']).toBe(testScalePointHeight);
      }

      if (isVertical && !testHeight) {
        expect(classViewScale['_scalePointsSize']).toBe(0);
      }

      if (!isVertical && testWidth) {
        expect(classViewScale['_scalePointsSize']).toBe(testScalePointWidth);
      }

      if (!isVertical && !testWidth) {
        expect(classViewScale['_scalePointsSize']).toBe(0);
      }

      classViewScale['_scalePointsSize'] = 0;
    },
  );

  test('Checking the "_setStyleForScale" method => "defaultSettings"', () => {
    classViewScale.update(testSettings);
    const { colorForScale } = testSettings;

    expect($elemScale.prop('style')).toHaveProperty('border-color', colorForScale);
    expect($elemScale.prop('style')).toHaveProperty('color', colorForScale);
    expect($elemScale.prop('style')).toHaveProperty('opacity', '1');
    expect($elemScale.prop('style')).toHaveProperty('pointer-events', '');
    expect($elemScale.html()).not.toMatch(/tabindex/);
  });

  test('Checking the "_setStyleForScale" method => option "colorForScale"', () => {
    testSettings.key = 'colorForScale';
    testSettings.colorForScale = 'red';
    classViewScale.update(testSettings);
    const { colorForScale } = testSettings;

    expect($elemScale.prop('style')).toHaveProperty('border-color', colorForScale);
    expect($elemScale.prop('style')).toHaveProperty('color', colorForScale);
  });

  test('Checking the "_setStyleForScale" method => option "showScale"', () => {
    testSettings.key = 'showScale';
    testSettings.showScale = false;
    classViewScale.update(testSettings);
    const $elemScalePoints = classViewScale['_$elemScalePoints'];

    expect($elemScalePoints.length > 0).toBeTruthy();

    $elemScalePoints.each((_, scalePoint) => {
      expect($(scalePoint).attr('tabindex')).toBe('-1');
    });
    expect($elemScale.prop('style')).toHaveProperty('opacity', '0');
    expect($elemScale.prop('style')).toHaveProperty('pointer-events', 'none');
  });

  test('Checking the "_checkingScaleSize" method => Main test with "defaultSettings" (if the number of scale points is odd and <= 6)', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_calcScalePointsSize');
    jest.spyOn<ViewScale, any>(classViewScale, '_setPropForSkipScalePoint');
    jest.spyOn<ViewScale, any>(classViewScale, '_checkingSkipScalePointSize');
    jest.spyOn(Map.prototype, 'set');

    const SCALE_POINTS_SIZE_PX = 200;
    classViewScale.update(testSettings);
    classViewScale['_scalePointsSize'] = SCALE_POINTS_SIZE_PX;

    expect(classViewScale['_skipScalePointsArray']).toHaveLength(0);

    classViewScale['_checkingScaleSize'](testSettings);
    const $elemScalePoints = classViewScale['_$elemScalePoints'];
    const sizeScalePointsArray = classViewScale['_$elemScalePoints'].length;
    const skipScalePointsArray = classViewScale['_skipScalePointsArray'];
    const mapSkipScalePoints = classViewScale['_mapSkipScalePoints'];
    const calcScalePointsSize = classViewScale['_calcScalePointsSize'];
    const checkingSkipScalePointSize = classViewScale['_checkingSkipScalePointSize'];

    $elemScalePoints.each((_, scalePoint) => {
      expect(classViewScale['_setPropForSkipScalePoint']).toHaveBeenCalledWith(scalePoint);
    });

    expect(sizeScalePointsArray > 0).toBeTruthy();
    expect(skipScalePointsArray.length > 0).toBeTruthy();

    skipScalePointsArray.forEach(($scalePointSkip) => {
      expect(calcScalePointsSize).toHaveBeenCalledWith($scalePointSkip[0], testSettings);
    });

    expect(mapSkipScalePoints.set).toHaveBeenCalledWith(expect.any(Number), [
      ...skipScalePointsArray,
    ]);

    expect(checkingSkipScalePointSize).toHaveBeenCalledWith(
      SLIDER_SIZE_PX,
      expect.any(Number),
      testSettings,
    );
  });

  test('Checking the "_checkingScaleSize" method => if the number of scale points is even and > 6"', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_calcScalePointsSize');
    jest.spyOn<ViewScale, any>(classViewScale, '_setPropForSkipScalePoint');
    testSettings.maxValue = 8;
    classViewScale.update(testSettings);
    const SCALE_POINTS_SIZE_PX = 200;

    classViewScale['_scalePointsSize'] = SCALE_POINTS_SIZE_PX;
    classViewScale['_checkingScaleSize'](testSettings);
    const $elemScalePoints = classViewScale['_$elemScalePoints'];
    const skipScalePointsArray = classViewScale['_skipScalePointsArray'];
    const calcScalePointsSize = classViewScale['_calcScalePointsSize'];

    $elemScalePoints.each((index, scalePoint) => {
      if (index % 2 !== 0) {
        expect(classViewScale['_setPropForSkipScalePoint']).toHaveBeenCalledWith(scalePoint);
      }
    });

    expect(skipScalePointsArray.length > 0).toBeTruthy();
    skipScalePointsArray.forEach(($scalePointSkip) => {
      expect(calcScalePointsSize).toHaveBeenCalledWith($scalePointSkip[0], testSettings);
    });
  });

  test('Checking the "_checkingScaleSize" method => if the number of scale points is odd, <= 6 and this is not the first or last value in the scale "', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_calcScalePointsSize');
    jest.spyOn<ViewScale, any>(classViewScale, '_setPropForSkipScalePoint');
    testSettings.maxValue = 6;
    classViewScale.update(testSettings);
    const SCALE_POINTS_SIZE_PX = 200;

    classViewScale['_scalePointsSize'] = SCALE_POINTS_SIZE_PX;
    classViewScale['_checkingScaleSize'](testSettings);
    const $elemScalePoints = classViewScale['_$elemScalePoints'];
    const skipScalePointsArray = classViewScale['_skipScalePointsArray'];
    const calcScalePointsSize = classViewScale['_calcScalePointsSize'];

    $elemScalePoints.each((index, scalePoint) => {
      const firstOrLastIndex = index === 0 || index === $elemScalePoints.length - 1;

      if (index % 2 === 0 && !firstOrLastIndex) {
        expect(classViewScale['_setPropForSkipScalePoint']).toHaveBeenCalledWith(scalePoint);
      }
    });

    expect(skipScalePointsArray.length > 0).toBeTruthy();
    skipScalePointsArray.forEach(($scalePointSkip) => {
      expect(calcScalePointsSize).toHaveBeenCalledWith($scalePointSkip[0], testSettings);
    });
  });

  test('Checking the "_setPropForSkipScalePoint" method', () => {
    classViewScale.update(testSettings);
    classViewScale['_skipScalePointsArray'].length = 0;
    const testScalePoint = classViewScale['_$elemScalePoints'][0];
    classViewScale['_setPropForSkipScalePoint'](testScalePoint);

    const $testScalePoint = $(testScalePoint);
    expect($testScalePoint[0].outerHTML).toMatchSnapshot();
    expect($testScalePoint.attr('class')).toMatch(/meta-slider__scale-point_skipped/);
    expect($testScalePoint.attr('tabindex')).toBe('-1');
    expect($testScalePoint.prop('style')).toHaveProperty('color', 'transparent');
    expect(classViewScale['_skipScalePointsArray'].length).toBe(1);
  });

  test('Checking the "_checkingSkipScalePointSize" method', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_calcScalePointsSize');
    jest.spyOn(Map.prototype, 'delete');
    classViewScale.update(testSettings);

    const CONTROL_SIZE_PX = 100;
    const MARGIN_PX = 100;
    const $testScalePoint = classViewScale['_$elemScalePoints'].eq(0);

    classViewScale['_setPropForSkipScalePoint']($testScalePoint[0]);
    classViewScale['_mapSkipScalePoints'].set(CONTROL_SIZE_PX, [$testScalePoint]);
    expect(classViewScale['_mapSkipScalePoints'].size > 0).toBeTruthy();

    classViewScale['_checkingSkipScalePointSize'](SLIDER_SIZE_PX, MARGIN_PX, testSettings);
    expect($testScalePoint[0].outerHTML).toMatchSnapshot();
    expect($testScalePoint.attr('tabindex')).toBeUndefined();
    expect($testScalePoint.attr('class')).not.toMatch(/meta-slider__scale-point_skipped/);
    expect(classViewScale['_calcScalePointsSize']).toHaveBeenCalledWith(
      $testScalePoint[0],
      testSettings,
    );
    expect(classViewScale['_mapSkipScalePoints'].delete).toHaveBeenCalledWith(CONTROL_SIZE_PX);
    expect(classViewScale['_mapSkipScalePoints'].size).toBe(0);
  });

  test('Checking the "_setEventResize" method, event "resize" => "defaultSettings" => init "_handleWindowResize"', () => {
    jest.spyOn<ViewScale, any>(classViewScale, '_checkingScaleSize');
    const mockMakeThrottlingHandler = jest.spyOn(makeThrottlingHandler, 'default');
    classViewScale.update(testSettings);

    jest.useFakeTimers();
    $(window).trigger(`resize.scale-${sliderID}`);

    jest.advanceTimersByTime(mockMakeThrottlingHandler.mock.calls[0][1]);

    expect(mockMakeThrottlingHandler).toHaveBeenCalled();
    expect(classViewScale['_checkingScaleSize']).toHaveBeenCalledWith(testSettings);
  });

  test('Checking the "_setEventResize" method => option "showScale"', () => {
    jest.spyOn($.fn, 'off');
    testSettings.key = 'showScale';
    testSettings.showScale = false;
    classViewScale.update(testSettings);

    expect($(window).off).toHaveBeenCalledWith(`resize.scale-${sliderID}`);
  });

  test('Checking the "_setEventResize" method => option "initScaleAdjustment"', () => {
    jest.spyOn($.fn, 'off');
    testSettings.key = 'initScaleAdjustment';
    testSettings.showScale = false;
    classViewScale.update(testSettings);

    expect($(window).off).toHaveBeenCalledWith(`resize.scale-${sliderID}`);
  });

  test('Checking the "_setEventsScalePoints" method, event "click" => init "_handleScalePointClick"', () => {
    jest.spyOn(classViewScale, 'notify').mockImplementation();

    classViewScale.update(testSettings);
    const $elemScalePoints = classViewScale['_$elemScalePoints'];

    expect($elemScalePoints.length > 0).toBeTruthy();

    $elemScalePoints.each((_, elemPoint) => {
      const $elemPoint = $(elemPoint);
      const eventClick = $.Event('click.scalePoint');
      eventClick.preventDefault = jest.fn();

      $elemPoint.trigger(eventClick);

      expect(eventClick.preventDefault).toHaveBeenCalled();
      expect(classViewScale.notify).toHaveBeenCalledWith(
        eventClick,
        Number($elemPoint.attr('data-value')),
      );
    });
  });
});
