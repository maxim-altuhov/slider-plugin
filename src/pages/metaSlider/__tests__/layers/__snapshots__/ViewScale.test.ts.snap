// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checking the "ViewScale" layer Checking the "_checkingSkipScalePointSize" method 1`] = `"<button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"left: 0%; color: transparent;\\" data-value=\\"1\\">1</button>"`;

exports[`Checking the "ViewScale" layer Checking the "_createScale" method => "defaultSettings" 1`] = `"<button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"left: 0%;\\" data-value=\\"1\\">1</button><button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"left: 25%;\\" data-value=\\"2\\">2</button><button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"left: 50%;\\" data-value=\\"3\\">3</button><button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"left: 75%;\\" data-value=\\"4\\">4</button><button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"left: 100%;\\" data-value=\\"5\\">5</button>"`;

exports[`Checking the "ViewScale" layer Checking the "_setPropForSkipScalePoint" method 1`] = `"<button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point meta-slider__scale-point_skipped\\" style=\\"left: 0%; color: transparent;\\" data-value=\\"1\\" tabindex=\\"-1\\">1</button>"`;

exports[`Checking the "ViewScale" layer, before first initialization. State before first initialization the "update" method 1`] = `
"<div class=\\"meta-slider js-meta-slider\\"><div class=\\"meta-slider__progress js-meta-slider__progress\\"></div>
    <button type=\\"button\\" class=\\"meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_left\\" data-value=\\"\\" data-text=\\"\\">
      <span class=\\"meta-slider__marker js-meta-slider__marker meta-slider__marker_left\\"></span>
    </button>
    <button type=\\"button\\" class=\\"meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_right\\" data-value=\\"\\" data-text=\\"\\">
      <span class=\\"meta-slider__marker js-meta-slider__marker meta-slider__marker_right\\"></span>
    </button>
    <div class=\\"meta-slider__scale js-meta-slider__scale\\">
      <button type=\\"button\\" class=\\"meta-slider__scale-point js-meta-slider__scale-point\\" style=\\"display: none;\\"></button>
    </div></div>"
`;
